src/neural/metal/mps/NetworkGraph.mm

```diff
@@ -0,0 +1,512 @@
+    [self setVariable:[NSString stringWithFormat:@"%@/add2", label] tensor:add2Tensor];
+    [self setVariable:[NSString stringWithFormat:@"%@/relu", label] tensor:reluTensor];
+
+    [self setVariable:label tensor:reluTensor];
+
+    return reluTensor;
+}
+
+-(nonnull MPSGraphTensor *) addPolicyMapLayerWithParent:(MPSGraphTensor * __nonnull)parent
+                                              policyMap:(uint32_t * __nonnull)policyMap
+                                                  label:(NSString *)label
+{
+    NSData * policyMapData = [NSData dataWithBytesNoCopy:policyMap
+                                                  length:kNumPolicyOutputs * sizeof(uint32_t)
+                                            freeWhenDone:NO];
+
+    MPSGraphTensor * mappingTensor = [_graph constantWithData:policyMapData
+                                                        shape:@[@(kNumPolicyOutputs)]
+                                                     dataType:MPSDataTypeUInt32];
+
+    MPSGraphTensor * flatConvTensor = [_graph reshapeTensor:parent
+                                                  withShape:@[parent.shape[0], @([parent sizeOfDimensions:@[@1, @2, @3]])]
+                                                       name:[NSString stringWithFormat:@"%@/flatten", label]];
+
+    MPSGraphTensor * policyTensor = [_graph gatherWithUpdatesTensor:flatConvTensor
+                                                      indicesTensor:mappingTensor
+                                                               axis:1
+                                                    batchDimensions:0
+                                                               name:[NSString stringWithFormat:@"%@/gather", label]];
+
+    [self setVariable:[NSString stringWithFormat:@"%@/constant", label] tensor:mappingTensor];
+    [self setVariable:[NSString stringWithFormat:@"%@/flatten", label] tensor:flatConvTensor];
+    [self setVariable:[NSString stringWithFormat:@"%@/gather", label] tensor:policyTensor];
+    [self setVariable:label tensor:policyTensor];
+
+    return policyTensor;
+}
+
+
+-(void) setVariable:(NSString * __nonnull)name
+             tensor:(MPSGraphTensor *)tensor
+{
+    if (![[_readVariables allKeys] containsObject:name]) return;
+
+    _readVariables[name] = tensor;
+}
+
+-(void) trackVariable:(NSString * __nonnull)name
+{
+    _readVariables[name] = [NSNull null];
+}
+
+-(void) dumpVariable:(NSString * __nonnull)name
+             batches:(NSUInteger)batches
+{
+    if (!_readVariables[name] || _readVariables[name] == [NSNull null]) {
+        NSLog(@"No variable '%@' found", name);
+        return;
+    }
+
+    MPSGraphTensor * variable = (MPSGraphTensor *) _readVariables[name];
+    NSUInteger size = [variable.shape[0] intValue] > 0 ? [variable size] : batches * [variable sizeOfDimensions:@[@1, @2, @3]];
+
+    if (variable.dataType == MPSDataTypeUInt32) {
+        uint32_t * dumpArray = (uint32_t *)malloc(size * sizeof(uint32_t));
```

Comment:
I'm not that familiar with Objective C, but doesn't this `malloc` need corresponding `free` somewhere?

Reply:
This is only used for debug and will likely be removed later.